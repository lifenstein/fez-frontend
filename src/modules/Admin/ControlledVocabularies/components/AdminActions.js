import React, { useState } from 'react';
import PropTypes from 'prop-types';
import IconButton from '@mui/material/IconButton';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import MoreVertIcon from '@mui/icons-material/MoreVert';
import { debounce } from 'throttle-debounce';

export const navigateToUrl = (uri, target, navigatedFrom, options) => {
    const fullUri = uri;
    window.open(fullUri, target, options);
};

export const AdminActions = ({ adminActions = [], vocab }) => {
    const [anchorEl, setAnchorEl] = useState(null);
    const open = Boolean(anchorEl);

    const cvoId = vocab;

    const handleClick = event => {
        setAnchorEl(event.currentTarget);
    };

    /* istanbul ignore next */
    const handleClose = () => {
        setAnchorEl(null);
    };

    const menuOptions = adminActions.map(action => {
        const linkTarget = action.inApp ? '_self' : '_blank';
        const options = action.options || null;
        const url = action.url?.(cvoId) ?? '';
        const clickHandler =
            action.onClick ??
            ((forceNewTab = false) =>
                debounce(
                    300,
                    event => {
                        navigateToUrl(url, event.ctrlKey || forceNewTab ? '_blank' : linkTarget, options);
                    },
                    { atBegin: true },
                ));
        const label = action.label;
        return {
            label,
            clickHandler,
        };
    });

    return (
        <React.Fragment>
            <IconButton
                id={`admin-actions-button-${cvoId}`}
                data-analyticsid={`admin-actions-button-${cvoId}`}
                data-testid={`admin-actions-button-${cvoId}`}
                aria-label="More"
                aria-haspopup="true"
                onClick={handleClick}
                size="large"
            >
                <MoreVertIcon fontSize="small" />
            </IconButton>
            <Menu
                id={`admin-actions-menu-${cvoId}`}
                data-testid={`admin-actions-menu-${cvoId}`}
                anchorEl={anchorEl}
                open={open}
                onClose={handleClose}
            >
                {menuOptions.map((option, index) => (
                    <MenuItem
                        key={index}
                        onClick={option.clickHandler}
                        onContextMenu={() => option.clickHandler(true)}
                        onAuxClick={() => option.clickHandler(true)}
                    >
                        {option.label}
                    </MenuItem>
                ))}
            </Menu>
        </React.Fragment>
    );
};

AdminActions.propTypes = {
    adminActions: PropTypes.array,
    vocab: PropTypes.number.isRequired,
};

export default React.memo(AdminActions);
