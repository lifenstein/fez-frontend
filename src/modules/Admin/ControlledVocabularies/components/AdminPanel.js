/* eslint-disable no-unused-vars */
import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm, getFormValues, getFormSyncErrors, Field, propTypes } from 'redux-form/immutable';
import Immutable from 'immutable';

import Button from '@mui/material/Button';
import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Checkbox from '@mui/material/Checkbox';
import CircularProgress from '@mui/material/CircularProgress';

import { useTheme } from '@mui/material/styles';
import useMediaQuery from '@mui/material/useMediaQuery';

import { Alert } from 'modules/SharedComponents/Toolbox/Alert';
import { StandardCard } from 'modules/SharedComponents/Toolbox/StandardCard';
import { TextField } from 'modules/SharedComponents/Toolbox/TextField';
import { ACTION } from '../ControlledVocabularyContext';

const rootId = 'update_dialog';

const renderCheckbox = ({ input, label, ...props }) => (
    <Checkbox label={label} checked={input.value ? true : false} onChange={input.onChange} {...props} />
);

const AdminPanel = ({
    action,
    locale,
    isOpen,
    title,
    id,
    hideActionButton = false,
    hideCancelButton = false,
    onAction,
    onCancelAction,
    onClose,
    noMinContentWidth,
    error,
    formValues,
    formErrors,
    handleSubmit,
    submitting,
    pristine,
    valid,
    parentId,
    ...props
}) => {
    const componentId = `${rootId}-${id}`;
    console.log(formValues.toJS(), formErrors, submitting);

    const theme = useTheme();
    const isMobileView = useMediaQuery(theme.breakpoints.only('xs')) || false;

    let containerStyles;
    if (action === ACTION.ADD) {
        containerStyles = !!parentId
            ? {}
            : {
                  backgroundColor: '#eee',
                  padding: '20px',
                  boxShadow: 'inset 0px 2px 4px 0px rgba(0,0,0,0.2)',
              };
    } else if (action === ACTION.EDIT) {
        containerStyles = {
            backgroundColor: '#eee',
            padding: '20px',
            boxShadow: 'inset 0px 2px 4px 0px rgba(0,0,0,0.2)',
        };
    }

    const _onCancelAction = () => {
        onClose?.();
        onCancelAction?.();
    };

    React.useEffect(() => {
        if (!props.initialized) props.initialize(props.initialValues);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [props.initialized]);

    return (
        <>
            {isOpen && (
                <Box
                    sx={{
                        marginBlockEnd: 2,
                        ...containerStyles,
                    }}
                >
                    <StandardCard title={title} standardCardId={`${componentId}`} subCard>
                        <form onSubmit={handleSubmit}>
                            <Box
                                id={`${componentId}-content`}
                                data-testid={`${componentId}-content`}
                                sx={{ minWidth: !noMinContentWidth ? 300 : 'auto', padding: 2 }}
                            >
                                <Grid container padding={0} spacing={2}>
                                    <Grid item xs={12}>
                                        <label htmlFor="cvo_title">Title</label>
                                        <Field
                                            component={TextField}
                                            variant="standard"
                                            required
                                            inputProps={{ maxLength: 255 }}
                                            fullWidth
                                            textFieldId="cvo-title"
                                            name="cvo_title"
                                            disabled={submitting}
                                        />
                                    </Grid>
                                    <Grid item xs={12}>
                                        <label htmlFor="cvo_desc">Description</label>
                                        <Field
                                            component={TextField}
                                            variant="standard"
                                            inputProps={{ maxLength: 255 }}
                                            fullWidth
                                            textFieldId="cvo-desc"
                                            name="cvo_desc"
                                            multiline
                                            minRows={2}
                                            disabled={submitting}
                                        />
                                    </Grid>
                                    <Grid item xs={12}>
                                        <label htmlFor="cvo_external_id" style={{ display: 'block' }}>
                                            External ID
                                        </label>
                                        <Field
                                            component={TextField}
                                            variant="standard"
                                            textFieldId="cvo-external-id"
                                            inputProps={{ maxLength: 10 }}
                                            name="cvo_external_id"
                                            disabled={submitting}
                                        />
                                    </Grid>
                                    <Grid item xs={12}>
                                        <label htmlFor="cvo_image_filename">Filename</label>
                                        <Field
                                            component={TextField}
                                            variant="standard"
                                            fullWidth
                                            textFieldId="cvo-image-filename"
                                            inputProps={{ maxLength: 64 }}
                                            name="cvo_image_filename"
                                            disabled={submitting}
                                        />
                                    </Grid>
                                    <Grid item xs={12}>
                                        <label htmlFor="cvo_order" style={{ display: 'block' }}>
                                            Order
                                        </label>
                                        <Field
                                            component={TextField}
                                            variant="standard"
                                            textFieldId="cvo-order"
                                            name="cvo_order"
                                            disabled={submitting}
                                        />
                                    </Grid>

                                    <Grid item xs={12}>
                                        <label htmlFor="cvo_hide" style={{ display: 'block' }}>
                                            Disabled
                                        </label>
                                        <Field
                                            component={renderCheckbox}
                                            variant="standard"
                                            id="cvo-hide"
                                            inputProps={{
                                                id: 'cvo-hide-input',
                                                'data-analyticsid': 'cvo-hide-input',
                                                'data-testid': 'cvo-hide-input',
                                            }}
                                            name="cvo_hide"
                                            type="checkbox"
                                            disabled={submitting}
                                        />
                                    </Grid>
                                </Grid>
                            </Box>
                            {(!hideCancelButton || !hideActionButton) && (
                                <Grid
                                    container
                                    id={`${rootId}-actions`}
                                    data-testid={`${rootId}-actions`}
                                    sx={{ marginTop: 2 }}
                                >
                                    <Grid item xs={12} justifyContent="flex-end" display={'flex'}>
                                        {!hideCancelButton && (
                                            <Button
                                                variant={'outlined'}
                                                onClick={_onCancelAction}
                                                id={`${rootId}-cancel-button`}
                                                data-testid={`${rootId}-cancel-button`}
                                                fullWidth={isMobileView}
                                                disabled={submitting}
                                                sx={{ marginInlineEnd: 2 }}
                                            >
                                                {locale.cancelButtonLabel}
                                            </Button>
                                        )}
                                        {!hideActionButton && (
                                            <Button
                                                variant="contained"
                                                autoFocus
                                                color={'primary'}
                                                type={'submit'}
                                                // onClick={_onAction}
                                                id={`${rootId}-action-button`}
                                                data-testid={`${rootId}-action-button`}
                                                fullWidth={isMobileView}
                                                disabled={pristine || submitting || !valid}
                                            >
                                                {submitting ? (
                                                    <CircularProgress
                                                        color="inherit"
                                                        size={25}
                                                        id={`${rootId}-progress`}
                                                        data-testid={`${rootId}-progress`}
                                                    />
                                                ) : (
                                                    locale.confirmButtonLabel
                                                )}
                                            </Button>
                                        )}
                                    </Grid>
                                </Grid>
                            )}
                            {!!error && (
                                <Grid container id={`${rootId}-alert`} data-testid={`${rootId}-alert`}>
                                    <Grid item xs={12}>
                                        <Alert title="Error" type="error_outline" message={error.message} />
                                    </Grid>
                                </Grid>
                            )}
                        </form>
                    </StandardCard>
                </Box>
            )}
        </>
    );
};

AdminPanel.propTypes = {
    ...propTypes, // all redux-form props
    action: PropTypes.oneOf(['', 'add', 'edit']).isRequired,
    locale: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
    row: PropTypes.object,
    isOpen: PropTypes.bool,
    noMinContentWidth: PropTypes.bool,
    hideActionButton: PropTypes.bool,
    hideCancelButton: PropTypes.bool,
    onAction: PropTypes.func,
    onCancelAction: PropTypes.func,
    onClose: PropTypes.func,
    props: PropTypes.object,
    alertProps: PropTypes.object,
    isBusy: PropTypes.bool,
};

export default React.memo(AdminPanel);
