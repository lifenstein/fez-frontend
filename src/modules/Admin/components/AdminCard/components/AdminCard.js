import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import { styled } from '@mui/material/styles';

const StyledRule = styled('hr')(({ theme }) => ({
    opacity: 0.1,
    border: 0,
    borderTop: `1px solid ${theme.palette.primary.main}`,
}));
const StyledTitle = styled(Typography)(({ theme }) => ({
    opacity: 0.85,
    color: theme.palette.primary.main,
    fontSize: theme.typography.fontSize,
    fontWeight: 500,
}));

export class Cards extends Component {
    static propTypes = {
        title: PropTypes.any,
        children: PropTypes.any,
    };
    render() {
        const { title, children } = this.props;
        return (
            <Grid container spacing={0} sx={{ mt: '8px' }} className={'AdminCard'}>
                <Grid item xs={12}>
                    <Grid item xs={12}>
                        <Grid container spacing={1}>
                            <Grid item xs>
                                <StyledRule />
                            </Grid>
                            <Grid item xs={'auto'} sx={{ ml: '-4px' }}>
                                <StyledTitle component={'h4'} color={'primary'}>
                                    &nbsp;{title}
                                </StyledTitle>
                            </Grid>
                        </Grid>
                    </Grid>
                </Grid>
                <Grid item xs={12}>
                    {children}
                </Grid>
            </Grid>
        );
    }
}

export const AdminCard = props => <Cards {...props} />;
export default AdminCard;
