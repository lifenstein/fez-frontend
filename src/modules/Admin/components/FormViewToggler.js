import React from 'react';
import { useTabbedContext } from 'context';

import Grid from '@material-ui/core/Grid';
import Tooltip from '@material-ui/core/Tooltip';
import Switch from '@material-ui/core/Switch';

import KeyboardIcon from '@material-ui/icons/Keyboard';
import { HelpIcon } from 'modules/SharedComponents/Toolbox/HelpDrawer';
import txt from 'locale/pages';

export const FormViewToggler = React.memo(() => {
    const { tabbed, toggleTabbed } = useTabbedContext();
    return (
        <Grid container direction="row" spacing={0} alignItems="center">
            <Grid item>
                <Tooltip title={`Switch to ${tabbed ? 'full form' : 'tabbed'} mode`}>
                    <Switch color="primary" checked={tabbed} onChange={toggleTabbed} value="tabbed" />
                </Tooltip>
            </Grid>
            <Grid item>
                <HelpIcon IconComponent={KeyboardIcon} {...txt.pages.edit.help} />
            </Grid>
        </Grid>
    );
});

export default FormViewToggler;
