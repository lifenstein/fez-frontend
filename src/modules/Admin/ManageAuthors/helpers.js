import { checkForExistingAuthor } from 'actions';
import { default as locale } from 'locale/components';

/* istanbul ignore next */
export const checkForExisting = (values, dispatch, props, field) =>
    dispatch(
        checkForExistingAuthor(
            values.get(field),
            field,
            values.get('aut_id'),
            locale.components.manageAuthors.editRow.validation,
            (!!props.asyncErrors && props.asyncErrors.toJS()) || {},
        ),
    );

const MAX_TOTAL = 100000;

export const composeAuthorAffiliationProblems = record => {
    const unique = record.fez_author_affiliation.reduce(
        (accumulated, current) =>
            accumulated.includes(current.af_author_id) ? accumulated : [...accumulated, current.af_author_id],
        [],
    );

    return record.fez_record_search_key_author_id
        .filter(author => author.rek_author_id !== 0)
        .map(author => {
            const isOrphaned = !unique.includes(author.rek_author_id);
            return {
                rek_author_id: author.rek_author_id,
                rek_author_id_lookup: author.rek_author_id_lookup,
                isOrphaned: isOrphaned,
                isIncomplete: !isOrphaned
                    ? record.fez_author_affiliation
                          .filter(item => item.af_author_id === author.rek_author_id)
                          .reduce((accumulated, current) => accumulated + current.af_percent_affiliation, 0) < MAX_TOTAL
                    : false,
            };
        })
        .filter(author => author.isOrphaned || author.isIncomplete);
};
