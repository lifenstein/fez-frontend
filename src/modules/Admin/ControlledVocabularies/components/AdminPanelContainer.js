import { connect } from 'react-redux';
import { reduxForm, getFormValues, getFormSyncErrors } from 'redux-form/immutable';
import Immutable from 'immutable';
import { bindActionCreators } from 'redux';

import * as actions from 'actions/viewControlledVocab';
import AdminPanel from './AdminPanel';

const FORM_NAME = 'ControlledVocabAdmin';

const rxWholeNumberOnly = new RegExp(/^\d+$/);

const validate = data => {
    const values = data.toJS();
    const errors = {};
    if (!values.cvo_title) {
        errors.cvo_title = 'Required';
    } else if (values.cvo_title.length > 255) {
        errors.cvo_title = 'Must be 255 characters or less';
    }
    if (!!values.cvo_desc && values.cvo_desc.length > 255) {
        errors.cvo_desc = 'Must be 255 characters or less';
    }

    if (typeof values.cvo_order !== 'undefined' && values.cvo_order !== '') {
        const num = Number(values.cvo_order);
        if (isNaN(num)) errors.cvo_order = 'Must be a number';
        else if (!rxWholeNumberOnly.test(values.cvo_order)) {
            errors.cvo_order = 'Must be a whole number above zero';
        }
    }

    return errors;
};

const mapStateToProps = (state, props) => {
    const formErrors = getFormSyncErrors(FORM_NAME)(state) || Immutable.Map({});
    console.log(props);
    return {
        onSubmit: props.onAction(props.parentId ?? null),
        formValues: getFormValues(FORM_NAME)(state) || Immutable.Map({}),
        formErrors: formErrors,
        disableSubmit: formErrors && !(formErrors instanceof Immutable.Map),
        initialValues: (state && state.get('vocabAdminReducer') && state.get('vocabAdminReducer').vocab) || {},
    };
};

const mapDispatchToProps = dispatch => ({
    actions: bindActionCreators(actions, dispatch),
});

const AdminPanelContainer = reduxForm({
    form: FORM_NAME,
    validate,
})(AdminPanel);

export default connect(mapStateToProps, mapDispatchToProps)(AdminPanelContainer);
