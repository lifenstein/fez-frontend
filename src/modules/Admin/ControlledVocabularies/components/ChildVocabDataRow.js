import React from 'react';
import PropTypes from 'prop-types';

import IconButton from '@mui/material/IconButton';
import Box from '@mui/material/Box';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
import Grid from '@mui/material/Grid';
import ChildVocabTable from './ChildVocabTable';
// import { useSelector, useDispatch } from 'react-redux';
// import * as actions from 'actions';

export const ChildVocabDataRow = ({ row }) => {
    // const dispatch = useDispatch();

    // const vocabOpened = useSelector(state => state.get('viewVocabReducer').vocabOpened);
    console.log('row=', row);
    const open = false;
    // const open = vocabOpened.indexOf(row.cvo_id) > -1;
    // const triggerChildren = openState => {
    //     // alert('here');
    //     // dispatch(actions.clearCCCollectionsList());
    //     console.log('openState=', openState);
    //     console.log('dispatch=', dispatch, 'actions=', typeof actions.setOpenedVocab);
    //     dispatch(actions.setOpenedVocab({ id: row.cvo_id, open: openState }));
    //     // if (openState) {
    //     //     dispatch(
    //     //         actions.loadChildVocabList({
    //     //             pid: row.cvo_id,
    //     //         }),
    //     //     );
    //     // }
    // };

    return (
        <Grid
            container
            key={row.cvo_id}
            data-testid={`row-${row.cvo_id}`}
            sx={{ boxSizing: 'border-box', boxShadow: '0 -1px 0 #eaeaea', padding: '15px 0px 0px' }}
        >
            <React.Fragment key={row.cvo_id}>
                <Grid container sx={{ paddingBottom: '10px' }}>
                    <Grid item xs={1} sm={1} md={1}>
                        <Box sx={{ float: 'left', width: '24px' }}>
                            <IconButton
                                sx={{ paddingTop: '5px' }}
                                aria-label="expand row"
                                size="small"
                                id={`expand-row-${row.cvo_id}`}
                                data-analyticsid={`expand-row-${row.cvo_id}`}
                                data-testid={`expand-row-${row.cvo_id}`}
                                // onClick={() => {
                                //     triggerChildren(!open);
                                // }}
                            >
                                {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
                            </IconButton>
                        </Box>
                    </Grid>
                    <Grid item md={8} sm={6} xs={6}>
                        <Box>{row.cvo_title}</Box>
                    </Grid>
                    <Grid item md={1} xs={2} sm={2}>
                        <Box>{/* row.cvo_image_filename*/}</Box>
                    </Grid>
                    <Grid item md={1} xs={2} sm={2}>
                        <Box>{row.cvo_external_id}</Box>
                    </Grid>
                    <Grid item md={1} xs={1} sm={1}>
                        {''}
                    </Grid>
                </Grid>
                {!!open && (
                    <Grid container>
                        <Grid item md={12}>
                            <ChildVocabTable parentRow={row} />
                        </Grid>
                    </Grid>
                )}
            </React.Fragment>
        </Grid>
    );
};
ChildVocabDataRow.propTypes = {
    conf: PropTypes.object,
    row: PropTypes.object,
    labels: PropTypes.object,
};
export default ChildVocabDataRow;
