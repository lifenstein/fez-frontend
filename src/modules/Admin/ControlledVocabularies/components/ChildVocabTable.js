/* eslint-disable no-unused-vars */
import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';

import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import Add from '@mui/icons-material/Add';
import TablePagination from '@mui/material/TablePagination';

import locale from 'locale/components';
import * as actions from 'actions';

import { InlineLoader } from 'modules/SharedComponents/Toolbox/Loaders';

import ChildVocabDataRow from './ChildVocabDataRow';
import { controlledVocabConfig } from 'config/controlledVocabConfig';
import { ControlledVocabulariesActionContext } from '../ControlledVocabularyContext';
import { ControlledVocabulariesStateContext } from '../ControlledVocabularyContext';
import Breadcrumbs from './Breadcrumbs';

const txt = locale.components.controlledVocabulary;
const labels = txt.columns.labels;

export const ChildVocabTable = ({ parentRow, locked }) => {
    const dispatch = useDispatch();
    const { onAdminAddActionClick } = useContext(ControlledVocabulariesActionContext);
    const state = useContext(ControlledVocabulariesStateContext);
    const { loadingChildVocab, childData, perPage, currentPage } = useSelector(state =>
        state.get('viewChildVocabReducer'),
    );

    React.useEffect(() => {
        const parentId = parentRow.cvo_id;

        dispatch(
            actions.loadChildVocabList({
                pid: parentId,
                rootId: parentId,
            }),
        );
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const handleAddActionClick = () => {
        const navs = document.querySelectorAll(`#vocabNav-${parentRow.cvo_id} [id^=nav]`);
        const parentId = +navs[navs.length - 1].id.replace('nav-', '');
        onAdminAddActionClick(parentId, parentRow.cvo_id);
    };

    // Event handler for button clicks
    const handleBreadcrumbClick = ({ id }) => {
        dispatch(
            actions.loadChildVocabList({
                pid: id,
                rootId: parentRow.cvo_id,
            }),
        );
    };

    const breadCrumbElements = childData[parentRow.cvo_id]?.path ?? [];
    if (!breadCrumbElements.find(em => em.id === parentRow.cvo_id)) {
        // add in parent node
        breadCrumbElements.unshift({ id: parentRow.cvo_id, title: parentRow.cvo_title });
    }

    const vocabList = childData[parentRow.cvo_id]?.data || [];
    const total = vocabList.length;

    let start = currentPage * perPage;
    let end = start + Math.min(perPage, total - perPage * currentPage);
    const handlePageChange = (event, value) => {
        dispatch(actions.setCurrentPage(value));
        start = value * perPage;
        end = start + Math.min(perPage, total - perPage * currentPage);
    };
    const handlePerPageChange = event => {
        const newPerPage = parseInt(event.target.value, 10);
        dispatch(actions.setVocabPerPage(newPerPage));
        dispatch(actions.setCurrentPage(0));
        start = 0;
        end = start + Math.min(newPerPage, total - newPerPage * currentPage);
    };

    return (
        <Box
            sx={{
                backgroundColor: '#eee',
                padding: '20px',
                boxShadow: 'inset 0px 2px 4px 0px rgba(0,0,0,0.2)',
            }}
            data-testid={`vocab-table-${parentRow.cvo_id}`}
            id={`vocab-table-${parentRow.cvo_id}`}
        >
            {!locked && (
                <Button
                    id={`admin-add-vocabulary-button-${parentRow.cvo_id}`}
                    data-testid={`admin-add-vocabulary-button-${parentRow.cvo_id}`}
                    startIcon={<Add />}
                    variant={'contained'}
                    color={'primary'}
                    sx={{ marginBottom: '10px' }}
                    onClick={handleAddActionClick}
                    disabled={state.isOpen}
                >
                    {txt.admin.addChildButtonLabel}
                </Button>
            )}

            <Box
                id={`portal-add-${parentRow.cvo_id}`}
                data-testid={`portal-add-${parentRow.cvo_id}`}
                sx={{ width: '100%' }}
            />
            <Box sx={{ minHeight: 200, backgroundColor: '#FFF', padding: '10px' }}>
                {loadingChildVocab[parentRow.cvo_id] && (
                    <Grid item md={12}>
                        <InlineLoader loaderId="childControlledVocab-page-loading" message={txt.loading.message} />
                    </Grid>
                )}
                {!!!loadingChildVocab[parentRow.cvo_id] && (childData[parentRow.cvo_id]?.data?.length ?? -1) >= 0 && (
                    <Grid container spacing={0}>
                        <Grid item md={12}>
                            <Breadcrumbs
                                id={`vocabNav-${parentRow.cvo_id}`}
                                data={breadCrumbElements}
                                onBreadcrumbClick={handleBreadcrumbClick}
                            />
                            <Typography
                                variant="body2"
                                sx={{ fontWeight: 600, marginBottom: '10px' }}
                                id={`total-vocab-${parentRow.cvo_id}`}
                                data-testid={`total-vocab-${parentRow.cvo_id}`}
                            >
                                {controlledVocabConfig.vocabCountTitle(
                                    childData[parentRow.cvo_id].data.length,
                                    parentRow.cvo_title,
                                )}{' '}
                            </Typography>
                        </Grid>
                        {/* Header Row */}
                        <Grid container spacing={0} sx={{ fontWeight: 400 }} data-testid="vocab-child-header">
                            <Grid item xs={12} sm={1}>
                                {labels.id}
                            </Grid>
                            <Grid item xs={12} sm={locked ? 5 : 4}>
                                {labels.title}
                            </Grid>
                            <Grid item xs={12} sm={5}>
                                {labels.desc}
                            </Grid>
                            <Grid item xs={12} sm={1}>
                                {labels.external_id}
                            </Grid>
                            {!locked && (
                                <Grid item xs={12} sm={1}>
                                    {labels.actions}
                                </Grid>
                            )}
                        </Grid>
                        {/* Data Row */}
                        <Grid container sx={{ paddingTop: '10px' }} data-testid="vocab-child-body">
                            {vocabList.slice(start, end).map(row => (
                                <ChildVocabDataRow
                                    key={row.controlled_vocab.cvo_id}
                                    row={row.controlled_vocab}
                                    parentId={row.cvr_parent_cvo_id}
                                    rootId={parentRow.cvo_id}
                                    locked={locked}
                                />
                            ))}
                            <Grid container spacing={0} direction="column" alignItems="center" justifyContent="center">
                                <Grid item xs={3}>
                                    {!!total && (
                                        <table>
                                            <tbody>
                                                <tr>
                                                    <TablePagination
                                                        data-testid="vocab-child-paging"
                                                        count={total}
                                                        page={currentPage}
                                                        rowsPerPage={perPage}
                                                        rowsPerPageOptions={[10, 25, 50, { label: 'All', value: -1 }]}
                                                        color="primary"
                                                        showFirstButton
                                                        showLastButton
                                                        onPageChange={handlePageChange}
                                                        onRowsPerPageChange={handlePerPageChange}
                                                    />
                                                </tr>
                                            </tbody>
                                        </table>
                                    )}
                                </Grid>
                            </Grid>
                        </Grid>
                    </Grid>
                )}
            </Box>
        </Box>
    );
};
ChildVocabTable.propTypes = {
    parentRow: PropTypes.object,
    locked: PropTypes.bool,
};
export default ChildVocabTable;
