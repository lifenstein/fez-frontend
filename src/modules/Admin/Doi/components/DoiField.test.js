import { shallow } from 'enzyme';
import { DoiField } from './DoiField';

import publicationTypeListResearchReport from 'mock/data/records/publicationTypeListResearchReport';
const record = publicationTypeListResearchReport.data[0];

jest.mock('modules/ViewRecord/components/AdditionalInformation');
jest.mock('react-html-parser');

import { formatPublicationDate } from 'modules/ViewRecord/components/AdditionalInformation';
import { parseHtmlToJSX } from 'helpers/general';

const setup = (testProps = {}, args = { isShallow: false }) => {
    const props = {
        classes: {},
        ...testProps,
    };

    return getElement(DoiField, props, args);
};

describe('DoiField', () => {
    it('should render default view', () => {
        const wrapper = setup({});
        // check for non existence of any elements beyond DoiField
        expect(wrapper.find('ForwardRef(Grid)[container="true"]').exists()).toBe(false);
    });

    it('should render unknown field value in readable format', () => {
        const wrapper = setup({ field: 'abc123', label: 'Unknown', data: 'testing' });
        expect(wrapper.find('span[data-testid="abc123"]').text()).toBe('"testing"');
    });

    it('should render known fields', () => {
        const testFn = ({ data, displayTypeLookup, field, label, test }) => {
            const wrapper = setup({
                data: data || record[field],
                displayTypeLookup,
                field,
                label: label || field,
            });
            test(wrapper);
        };

        const knownFields = [
            // Author
            {
                field: 'fez_record_search_key_author',
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-author-0"]').text()).toBe(
                        record.fez_record_search_key_author[0].rek_author,
                    );
                },
            },
            // HTML
            {
                field: 'rek_description',
                test: () => {
                    expect(parseHtmlToJSX).toHaveBeenCalledWith(record.rek_description);
                },
            },
            // Date
            {
                field: 'rek_date',
                test: () => {
                    expect(formatPublicationDate).toHaveBeenCalledWith(record.rek_date, undefined);
                },
            },
            // Edition
            {
                field: 'fez_record_search_key_edition',
                data: {
                    rek_edition: '1',
                },
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-edition"]').text()).toBe('1');
                },
            },
            // Example of list keys
            {
                field: 'fez_record_search_key_isbn',
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-isbn"]').text()).toBe(
                        record.fez_record_search_key_isbn[0].rek_isbn,
                    );
                },
            },
            // Example of single entry keys
            {
                field: 'fez_record_search_key_publisher',
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-publisher"]').text()).toBe(
                        record.fez_record_search_key_publisher.rek_publisher,
                    );
                },
            },
            // Example of generic entry with label
            {
                field: 'fez_record_search_key_conference_location',
                data: {
                    rek_conference_location: 'test location',
                },
                label: 'Conference location',
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-conference-location"]').text()).toBe('test location');
                    expect(wrapper.find('span[data-testid="rek-conference-location-label"]').text()).toBe(
                        'Conference location',
                    );
                },
            },

            /* Special Cases */

            // Author with ORCID
            {
                field: 'fez_record_search_key_author',
                data: [
                    {
                        rek_author_id: 10101,
                        rek_author: 'First Last',
                        aut_orcid_id: '101010-1010101',
                    },
                ],
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-author-0"]').text()).toBe('First Last');
                    expect(wrapper.find('ExternalLink[data-testid="rek-author-0-orcid-link"]').props().href).toBe(
                        'https://orcid.org/101010-1010101',
                    );
                },
            },
            // Empty contributors list
            {
                field: 'fez_record_search_key_contributor',
                data: [],
                test: wrapper => {
                    // check for non existence of any elements beyond DoiField
                    expect(wrapper.find('ForwardRef(Grid)[container="true"]').exists()).toBe(false);
                },
            },
            // Empty list item
            {
                field: 'fez_record_search_key_isbn',
                data: [],
                test: wrapper => {
                    // check for non existence of any elements beyond DoiField
                    expect(wrapper.find('ForwardRef(Grid)[container="true"]').exists()).toBe(false);
                },
            },
            // Date with custom format
            {
                field: 'rek_date',
                displayTypeLookup: 'YYYY',
                test: () => {
                    expect(formatPublicationDate).toHaveBeenCalledWith(record.rek_date, 'YYYY');
                },
            },
            // Depositor name
            {
                field: 'rek_author-name',
                data: 'Test Depositor',
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-author-name"]').text()).toBe('Test Depositor');
                },
            },
            // Depositor email
            {
                field: 'rek_author-email',
                data: 'example@uq.edu.au',
                test: wrapper => {
                    const entry = wrapper.find('span[data-testid="rek-author-email"]');
                    expect(entry.text()).toBe('example@uq.edu.au');
                    const renderedEmail = shallow(entry.props().children);
                    expect(renderedEmail.props().href).toBe('mailto:example@uq.edu.au');
                },
            },
            // Series with ": no"
            {
                field: 'fez_record_search_key_series',
                data: {
                    rek_series: 'Book series: no.1',
                },
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-series"]').text()).toBe('Book series');
                },
            },
            // Series with ";"
            {
                field: 'fez_record_search_key_series',
                data: {
                    rek_series: 'Book series 1; Book series 2',
                },
                test: wrapper => {
                    expect(wrapper.find('span[data-testid="rek-series"]').text()).toBe('Book series 1');
                },
            },
            // Non-numeric Edition
            {
                field: 'fez_record_search_key_edition',
                data: {
                    rek_edition: '15th',
                },
                test: wrapper => {
                    expect(toJson(wrapper.find('span[data-testid="rek-edition"]'))).toBe(null);
                },
            },
        ];

        knownFields.map(testFn);
    });
});
