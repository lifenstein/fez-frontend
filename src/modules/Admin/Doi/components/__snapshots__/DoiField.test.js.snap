// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoiField should render default view 1`] = `<div />`;

exports[`DoiField should render known fields 1`] = `<div />`;

exports[`DoiField should render known fields 2`] = `<div />`;
